tests/shouldwork/AutoReg/AutoReg.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=10 #-}
+-- Plugins don't load automatically when running the output tests
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.KnownNat.Solver #-}
+
+module AutoReg where
+import Data.Int
+import Clash.Prelude
+import Clash.Class.AutoReg (AutoReg)
+import Clash.Sized.Internal.BitVector
+import Control.Monad (when)
+import qualified Data.List as L
+import System.Environment (getArgs)
+import System.FilePath ((</>), takeDirectory)
+
+data Tup2 a b = MkTup2 { getA :: a, getB :: b } deriving (Generic,NFDataX)
+instance (BitPack a,BitPack b, KnownNat (BitSize a),KnownNat (BitSize b)) => BitPack (Tup2 a b)
+deriveAutoReg ''Tup2
+
+
+data Tup3 a b c = MkTup3 { fieldA :: a, fieldB :: b, fieldC :: c } deriving (Generic,NFDataX)
+instance
+  ( BitPack a,BitPack b,BitPack c
+  , KnownNat (BitSize a), KnownNat (BitSize b), KnownNat (BitSize c)
+  )
+    => BitPack (Tup3 a b c)
+deriveAutoReg ''Tup3
+
+
+newtype OtherPair a b = OtherPair (Tup2 a b) deriving (Generic,NFDataX)
+instance (BitPack a,BitPack b, KnownNat (BitSize a),KnownNat (BitSize b)) => BitPack (OtherPair a b)
+deriveAutoReg ''OtherPair
+
+data Tup2_ a b c = MkTup2_ a b deriving (Generic,NFDataX)
+instance (BitPack a,BitPack b, KnownNat (BitSize a),KnownNat (BitSize b)) => BitPack (Tup2_ a b c)
+deriveAutoReg ''Tup2_
+
+data Concrete = BoolAndInt Bool Int8 deriving (Generic,NFDataX,BitPack)
+deriveAutoReg ''Concrete
+
+data InfixDataCon a b = a :-.- b deriving (Generic,NFDataX)
+instance (BitPack a,BitPack b,KnownNat (BitSize a),KnownNat (BitSize b)) => BitPack (InfixDataCon a b)
+deriveAutoReg ''InfixDataCon
+
+
+test
+  :: forall a dom n rest
+   . ( HiddenClockResetEnable dom
+     , AutoReg a, BitPack a
+     , KnownNat (BitSize a), KnownNat n, KnownNat rest
+     , rest ~ (n-(BitSize a))
+     )
+  => Signal dom (BitVector n) -> Signal dom a
+test = autoReg (unpack 0) . fmap (unpack . fst . split# @rest)
```

Comment:
Maybe we should add a test for `autoReg1` too, confirming that it only creates to top-level registers.

Reply:
We decided to remove `autoReg1` for now